### PR TITLE
Assignment 04

### PR DESCRIPTION
**Pull request by lhwitten**

### CODE REVIEW

File: mini-projects/00-jq-rs/Cargo.toml [hunk 1]
```diff
@@ -4,3 +4,11 @@ version = "0.1.0"
...
 
 [dependencies]
+ansi_term = "0.12.1"
+anyhow = "1.0.89"
+assert_cmd = "2.0.16"
+clap = { version = "4.5.16", features = ["derive"] }
+colored = "2.1.0"
+predicates = "3.1.2"
+serde = "1.0.210"
+serde_json = "1.0.128"
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/Cargo.toml:14]: as a quick note, you can use the `preserve_order` feature of serve to maintain the original json order

File: mini-projects/00-jq-rs/src/main.rs [hunk 2]
```diff
@@ -1,3 +1,99 @@
...
+    // Apply the filters
+    let initial_values = vec![json_obj];
+    let output_values = match filters::apply_filters(initial_values, &filters) {
+        Ok(values) => values,
+        Err(e) => {
+            eprintln!("Error applying filters: {}", e);
+            std::process::exit(1);
+        }
+    };
+    let monochrome_output = args.monochrome_output.unwrap_or(false);
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:81]: might have been easier to just specify default values

File: mini-projects/00-jq-rs/src/main.rs [hunk 3]
```diff
@@ -1,3 +1,99 @@
...
+    json_file: Option<PathBuf>,
+}
+
 fn main() {
     println!("Hello World!");
+    let args = Args::parse();
+    //println!("{:?}", args);
+
+    // println!("{:?}", args);
+    //let input_vec = input::read_general_input(args.file)?;
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:47]: nit: delete stale code

File: mini-projects/00-jq-rs/src/main.rs [hunk 4]
```diff
@@ -1,3 +1,99 @@
...
+    #[clap(short, long)]
+    indent: Option<usize>,
+
+    filter_string: String,
+
+    json_file: Option<PathBuf>,
+}
+
 fn main() {
     println!("Hello World!");
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:42]: this annoyed me as I was running the filters

File: mini-projects/00-jq-rs/src/input.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+pub trait input {
+    fn get_lines(&self) -> Result<Vec<String>>;
+}
+
+pub struct file_input {
+    path: PathBuf,
+}
+
+impl input for file_input {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/input.rs:14]: nice use of traits, even if you just need one way of implementing for now

Comment (ayushchakra) [reply mini-projects/00-jq-rs/src/input.rs:14]: I see this didn't actually end up getting used..... still nice thought though! curious why you decided to pivot away from this

File: mini-projects/00-jq-rs/src/main.rs [hunk 6]
```diff
@@ -1,3 +1,99 @@
...
+    let json_obj: serde_json::Value = serde_json::from_reader(reader).unwrap();
+
+    // if let Some(array) = json_obj.as_array() {
+
+    // }
+
+    let filters = match filters::parse_filter_sequence(&args.filter_string) {
+        Ok(f) => f,
+        Err(e) => {
+            eprintln!("Error parsing filter: {}", e);
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:67]: Love the error handling here

File: mini-projects/00-jq-rs/src/filters.rs [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+#[derive(Debug, PartialEq)]
+pub enum Filter {
+    Identity,
+    ObjectIdentifier(String),
+    ArrayIndex(isize),
+    ArraySlice(Option<isize>, Option<isize>),
+    ArrayIterator,
+    Add,
+    Length,
+    Del(Box<Filter>),
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:18]: This is an excellent way to define the filters, especially for the Del to leverage information you already have

Comment (ayushchakra) [reply mini-projects/00-jq-rs/src/filters.rs:18]: huge fan of this clean interface btw

File: mini-projects/00-jq-rs/src/main.rs [hunk 8]
```diff
@@ -1,3 +1,99 @@
...
+            std::process::exit(1);
+        }
+    };
+
+    // Apply the filters
+    let initial_values = vec![json_obj];
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:73]: I see you're wrapping the json in a vec to make the array iterator interface more generic later - I see what you're doing, but it *may* have been better to just have the value within apply_filters to be an enum of either a actual SerdeJson or an Array of SerdeJsons, that would make it more explicit what the distinction is

File: mini-projects/00-jq-rs/src/filters.rs [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+            let result = apply_single_filter(value, filter)?;
+            next_values.extend(result);
+        }
+        current_values = next_values;
+    }
+
+    Ok(current_values)
+}
+
+pub fn apply_single_filter(value: Value, filter: &Filter) -> Result<Vec<Value>, String> {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:50]: this function would have benefitted from utilizing more helper functions

File: mini-projects/00-jq-rs/src/filters.rs [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    match filter {
+        Filter::Identity => Ok(vec![value]),
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:52]: similar to my previous comment about how you handle the array iterators, this line to me seems silly to need to wrap the value in a vec, just for you to call extend it in the higher function

File: mini-projects/00-jq-rs/src/filters.rs (hunk 9 above)
Comment (ayushchakra) [reply mini-projects/00-jq-rs/src/filters.rs:50]: but love the generic interface of just taking in the filter enum

File: mini-projects/00-jq-rs/src/filters.rs [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+                Err("Value is not an array.".to_string())
+            }
+        }
+        Filter::ArraySlice(start_opt, end_opt) => {
+            if let Value::Array(array) = value {
+                let len = array.len() as isize;
+                let start = start_opt.unwrap_or(0);
+                let end = end_opt.unwrap_or(len);
+                let start = if start < 0 { len + start } else { start };
+                let end = if end < 0 { len + end } else { end };
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filters.rs:85]: took me a hot sec, but I see what you're doing here - I almost feel like it would have been better if you just inlined each of these for each unique variable

File: mini-projects/00-jq-rs/src/printers.rs [hunk 12]
```diff
@@ -0,0 +1,243 @@
+use ansi_term::Colour;
+use serde::Serialize;
+use serde_json::Value;
+
+pub struct Formatter {
+    pub color_output: bool,
+    pub indent: usize,
+    pub compact: bool,
+    pub sort_keys: bool,
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/printers.rs:9]: Rather than having multiple parameters within this one formatting struct, you could have split this into more dedicate formatters (like a ColorFormatter) to avoid excessive if statements in the `print_value` function

**Comment by ayushchakra** [comment]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 4/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

#### Comments
(-1.5) Comments are not correctly bubbled up

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

#### Comments
Identity: Correct implementation and unit test
Object-Identifier: Correct implementation and unit test
Array Index: Correct implementation and unit test
Array Slice: Correct implementation and unit test
(-1) Pipe: Correct implementation, unit test exists for detecting a pipe, but would have been good to explicitly check that the filter chains work as expected
Array Iterator: Correct implementation and unit tests

(-2) Not taking points off everywhere, but as a whole, you test the positive cases for your filters quite well, but don't test for when invalid inputs are given.

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

#### Comments
Add: Correct implementation and unit test
Length: Correct implementation and unit test
Del: Correct implementation and unit test

(-2) Similar to above, would have been good to have a slighty wider array of tests. This is especially important in this section for stuff like add since it can be defined differently for different datatypes.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

#### Comments
(-1) Color Printing: Color printing is implemented, but the `JQ_COLORS` env var is not read for the user to customize these formats. 
Monochrome Printing: Correctly implemented and unit tests
Sorted Keys: Correctly implemented and unit tests
Indents: Correctly implemented and unit tests
Compact Output: Correctly implemented and unit tests

(-2) Effort was made for testing the output formats, but more explicit assertions would make them more effective tests (could be accomplished by faking out what you are writing to in order to consume the desired output)

### Points
**Features Score: 53/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 12/15**

#### Comments
Scalability: 5/5 - Overall, the scalability of your application was very strong, nice work.
General Cleanliness: 2/4 - taking some points off her for the dead code throughout the assignment. Could have also benefitted from some slightly more verbose commenting
-1 for clippy errors

## Summary
Add up the point totals here

**Total Score: 69/80**
